README.md
# WaiverDB

![logo of WaiverDB](https://pagure.io/waiverdb/raw/master/f/logo.png)

## What is WaiverDB

WaiverDB is a companion service to
[ResultsDB](https://pagure.io/taskotron/resultsdb), for recording waivers
against test results.

## Quick development setup

Install dependencies:

    $ sudo dnf builddep waiverdb.spec

Configure Postgres on the local machine, with a `waiverdb` database:

    $ sudo dnf install postgresql-server
    $ sudo postgresql-setup --initdb
    $ sudo systemctl enable --now postgresql
    $ sudo -u postgres createuser --superuser $USER
    $ createdb waiverdb

Create a local configuration file:

    $ cp conf/settings.py.example conf/settings.py

Populate the database:

    $ PYTHONPATH=. DEV=true python waiverdb/manage.py db upgrade

Run the server:

    $ PYTHONPATH=. DEV=true python waiverdb/manage.py run -h localhost -p 5004 --debugger

The server is now running at <http://localhost:5004> and API calls can be sent to
<http://localhost:5004/api/v1.0>. All data is stored in the `waiverdb` Postgres 
database on the local machine. You can verify the server is running correctly 
by visiting <http://localhost:5004/api/v1.0/about>.


## Adjusting configuration

You can configure this app by copying `conf/settings.py.example` into
`conf/settings.py` and adjusting values as you see fit. It overrides default
values in `waiverdb/config.py`.

## Running test suite

You can run this test suite with the following command::

    $ py.test tests/

The test suite will drop and re-create a Postgres database named 
`waiverdb_test`. By default, it expects to have superuser access to Postgres on 
the local machine.

To test against all supported versions of Python, you can use tox::

    $ sudo dnf install python3-tox
    $ tox

## Building the docs

You can view the docs locally with::

    $ cd docs
    $ make html
    $ firefox _build/html/index.html

## Viewing published fedmsgs

You can view fedmsgs published when new waivers get created by doing::

    $ fedmsg-relay --config-filename fedmsg.d/config.py &
    $ fedmsg-tail --config fedmsg.d/config.py --no-validate --really-pretty

### WaiverDB CLI
WaiverDB has a command-line client interface for creating new waivers against test
results. A sample configuration is installed as ``/usr/share/doc/waiverdb/client.conf.example``.
Copy it to ``/etc/waiverdb/client.conf`` and edit it there. Or you can use ``--config-file``
to specify one.
```
Usage: waiverdb-cli [OPTIONS]

  Creates new waivers against test results.

  Examples:

      waiverdb-cli -r 123 -r 456 -p "fedora-26" -c "It's dead!"
or

      waiverdb-cli -t dist.rpmlint -s '{"item": "python-requests-1.2.3-1.fc26", "type": "koji_build"}' -p "fedora-26" -c "It's dead!"


Options:
  -C, --config-file PATH      Specify a config file to use
  -r, --result-id INTEGER     Specify one or more results to be waived
  -s, --subject TEXT          Specify one subject for a result to waive
  -t, --testcase TEXT         Specify a testcase for the subject
  -p, --product-version TEXT  Specify one of PDC's product version
                              identifiers.
  --waived / --no-waived      Whether or not the result is waived
  -c, --comment TEXT          A comment explaining why the result is waived
  -h, --help                  Show this message and exit.
```


docs/release-notes.rst
=============
Release Notes
=============

WaiverDB 0.9
============

Released 1 Mar 2018.

*  The usage of ``JSONB`` has been replaced with the older ``JSON`` column
   type, in order to maintain compatibility with PostgreSQL 9.2 on RHEL7
   (#134).

WaiverDB 0.8
============

Released 16 Feb 2018.

* Removed support to SQLite in favor of PostgreSQL.

* Fixed database migration to use the correct column type for the
  ``waiver.subject`` column (#129).

* Added information on the README file on how to configure the db.

WaiverDB 0.7
============

Released 16 Feb 2018.

* Fixed the database migration strategy for Openshift deployment (#121).
  The migration step is now run in a pre-deployment hook. Previously it ran in
  a post-start pod hook which did not work correctly in some situations.

WaiverDB 0.6
============

Released 13 Feb 2018.

* Dummy authentication for CLI for developing and debugging reasons.

* Added logo in the README page.

* You can now waive the absence of a result. Now it is possible to 
  submit waivers using a subject/testcase.

* Backward compatibility for submitting a waiver using the result_id.
  This feature will be removed in the near future.

WaiverDB 0.5
============

Released 17 Jan 2018.

* Database migrations have been introduced, and will be a part of future
  releases.  Users upgrading to 0.5 will need to run these commands::

  $ waiverdb db stamp 0a27a8ad723a
  $ waiverdb db upgrade

* Error messages are now returned by the API in JSON format.

* A new authentication method: ssl auth.  See the docs for more on
  configuration.

* The API now supports a proxyuser argument.  A limited set of superusers,
  configured server-side, are able to submit waivers on behalf of other users.

WaiverDB 0.4
============

Released 08 Nov 2017.

A number of issues have been resolved in this release:

* New WaiverDB CLI for creating waivers (#82).

* New `/about` API endpoint to expose the current running version and the method
  used for authentication of the server.

* Improved the process of building docs by using sphinxcontrib.issuetracker
  extension.

WaiverDB 0.3
============

Released 26 Sep 2017.

A number of issues have been resolved in this release:

* Fixed some type errors in the API docs examples (#73).

* Updated README to recommend installing package dependencies using dnf builddep (#74).

* Fixed the health check API to return a proper error if the application is not
  able to serve requests (#75).

Other updates:

* Supports a new HTTP API `/api/v1.0/waivers/+by-result-ids`.
* Package dependencies are switched to python2-* packages in Fedora.

WaiverDB 0.2
============

Released 16 June 2017.

* Supports containerized deployment in OpenShift. ``DATABASE_PASSWORD`` and
  ``FLASK_SECRET_KEY`` can now be passed in as environment variables instead of
  being defined in the configuration file.

* Supports publishing messages over AMQP, in addition to Fedmsg.
  The ``ZEROMQ_PUBLISH`` configuration option has been renamed to
  ``MESSAGE_BUS_PUBLISH``.

* The :file:`/etc/waiverdb/settings.py` configuration file is no longer
  installed by default. For new installations, you can copy the example
  configuration from :file:`/usr/share/doc/waiverdb/conf/settings.py.example`.

* Numerous improvements to the test and build process for WaiverDB.

WaiverDB 0.1
============

Initial release, 12 April 2017.


waiverdb/cli.py
# SPDX-License-Identifier: GPL-2.0+
"""
CLI for creating new waivers against test results.
"""
import click
import requests
import json
import configparser

requests_session = requests.Session()


def validate_config(config):
    """
    Validates the configuration needed for WaiverDB
    :return: None or ClickException
    """
    # Verify that all necessary config options are set
    config_error = ('The config option "{0}" is required')
    if not config.has_option('waiverdb', 'auth_method'):
        raise click.ClickException(config_error.format('auth_method'))
    auth_method = config.get('waiverdb', 'auth_method')
    if auth_method not in ['OIDC', 'Kerberos', 'dummy']:
        raise click.ClickException('The WaiverDB authentication mechanism of '
                                   '"{0}" is not supported'.format(auth_method))
    required_configs = ['api_url']
    if auth_method == 'OIDC':
        required_configs.append('oidc_id_provider')
        required_configs.append('oidc_client_id')
        required_configs.append('oidc_scopes')
    for required_config in required_configs:
        if not config.has_option('waiverdb', required_config):
            raise click.ClickException(config_error.format(required_config))
    if not config.has_option('waiverdb', 'resultsdb_api_url'):
        raise click.ClickException(config_error.format('resultsdb_api_url'))


def check_response(resp, data, result_id=None):
    if result_id:
        msg = 'for result with id {0}'.format(result_id)
    else:
        msg = 'for result with subject {0} and testcase {1}'.format(json.dumps(data['subject']),
                                                                    data['testcase'])
    if not resp.ok:
        try:
            error_msg = resp.json()['message']
        except (ValueError, KeyError):
            error_msg = resp.text
        raise click.ClickException(
            'Failed to create waiver {0}:\n{1}'
            .format(msg, error_msg))
    click.echo('Created waiver {0} {1}'.format(
        resp.json()['id'], msg))


@click.command(context_settings={'help_option_names': ['-h', '--help']})
@click.option('--config-file', '-C', default='/etc/waiverdb/client.conf',
              type=click.Path(exists=True),
              help='Specify a config file to use')
@click.option('--result-id', '-r', multiple=True, type=int,
              help='Specify one or more results to be waived')
@click.option('--subject', '-s',
              help='Specify one subject for a result to waive')
@click.option('--testcase', '-t',
              help='Specify a testcase for the subject')
@click.option('--product-version', '-p',
              help='Specify one of PDC\'s product version identifiers.')
@click.option('--waived/--no-waived', default=True,
              help='Whether or not the result is waived')
@click.option('--comment', '-c',
              help='A comment explaining why the result is waived')
def cli(comment, waived, product_version, testcase, subject, result_id, config_file):
    """
    Creates new waiver against test results.

    Examples:

        waiverdb-cli -r 123 -r 456 -p "fedora-26" -c "It's dead!"

        or

        waiverdb-cli -t dist.rpmlint -s '{"item": "python-requests-1.2.3-1.fc26",
                                          "type": "koji_build"}'
                     -p "fedora-26" -c "It's dead!"

    """
    config = configparser.SafeConfigParser()

    config.read(config_file)
    validate_config(config)

    result_ids = result_id
    if not product_version:
        raise click.ClickException('Please specify product version')
    if result_ids and (subject or testcase):
        raise click.ClickException('Please specify result_id or subject/testcase. Not both')
    if not result_ids and not subject:
        raise click.ClickException('Please specify one subject')
    if not result_ids and not testcase:
        raise click.ClickException('Please specify testcase')

    auth_method = config.get('waiverdb', 'auth_method')
    data_list = []
    if not result_ids:
        data_list.append({
            'subject': json.loads(subject),
            'testcase': testcase,
            'waived': waived,
            'product_version': product_version,
            'comment': comment
        })

    # XXX - TODO - remove this in a future release.  (for backwards compat)
    for result_id in result_ids:
        data_list.append({
            'result_id': result_id,
            'waived': waived,
            'product_version': product_version,
            'comment': comment
        })

    api_url = config.get('waiverdb', 'api_url')
    if auth_method == 'OIDC':
        # Try to import this now so the user gets immediate feedback if
        # it isn't installed
        try:
            import openidc_client  # noqa: F401
        except ImportError:
            raise click.ClickException('python-openidc-client needs to be installed')
        # Get the auth token using the OpenID client.
        oidc_client_secret = None
        if config.has_option('waiverdb', 'oidc_client_secret'):
            oidc_client_secret = config.get('waiverdb', 'oidc_client_secret')
        oidc = openidc_client.OpenIDCClient(
            'waiverdb',
            config.get('waiverdb', 'oidc_id_provider'),
            {'Token': 'Token', 'Authorization': 'Authorization'},
            config.get('waiverdb', 'oidc_client_id'),
            oidc_client_secret)
        scopes = config.get('waiverdb', 'oidc_scopes').strip().splitlines()

        for data in data_list:
            resp = oidc.send_request(
                scopes=scopes,
                url='{0}/waivers/'.format(api_url.rstrip('/')),
                data=json.dumps(data),
                headers={'Content-Type': 'application/json'},
                timeout=60)
            check_response(resp, data, data.get('result_id', None))
    elif auth_method == 'Kerberos':
        # Try to import this now so the user gets immediate feedback if
        # it isn't installed
        try:
            import requests_gssapi  # noqa: F401
        except ImportError:
            raise click.ClickException(
                'python-requests-gssapi needs to be installed')
        auth = requests_gssapi.HTTPKerberosAuth(
            mutual_authentication=requests_gssapi.OPTIONAL)
        for data in data_list:
            resp = requests.request('POST', '{0}/waivers/'.format(api_url.rstrip('/')),
                                    data=json.dumps(data), auth=auth,
                                    headers={'Content-Type': 'application/json'},
                                    timeout=60)
            if resp.status_code == 401:
                raise click.ClickException('WaiverDB authentication using GSSAPI failed. '
                                           'Make sure you have a valid Kerberos ticket.')
            check_response(resp, data, data.get('result_id', None))
    elif auth_method == 'dummy':
        for data in data_list:
            resp = requests.request('POST', '{0}/waivers/'.format(api_url.rstrip('/')),
                                    data=json.dumps(data), auth=('user', 'pass'),
                                    headers={'Content-Type': 'application/json'},
                                    timeout=60)
            check_response(resp, data, data.get('result_id', None))


if __name__ == '__main__':
    cli()  # pylint: disable=E1120
